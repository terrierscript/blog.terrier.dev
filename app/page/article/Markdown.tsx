import React, { FC } from "react"
import ReactMarkdown from "react-markdown"
import { Prism as SyntaxHighlighter } from "react-syntax-highlighter"
import style from "react-syntax-highlighter/dist/cjs/styles/prism/tomorrow"
import { nl2brRemark } from "./unified/nl2br"

export const CodeBlock = ({ value, language }) => {
  return (
    <SyntaxHighlighter language={language} style={style}>
      {value}
    </SyntaxHighlighter>
  )
}

export const RawMarkdown: FC<{ markdown: string }> = ({ markdown }) => {
  return (
    <ReactMarkdown
      source={markdown}
      escapeHtml={false} // for web components
      renderers={{ code: CodeBlock }}
      plugins={[nl2brRemark]}
    />
  )
}
