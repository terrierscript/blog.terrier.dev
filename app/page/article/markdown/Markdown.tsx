import React, { FC } from "react"
import ReactMarkdown from "react-markdown/with-html" // for web components
import { nl2brRemark } from "./unified/nl2br"
import styled from "@emotion/styled"
import { PrismCodeBlock } from "./CodeBlock"
import {
  ARTICLE_COLOR,
  // TITLE_COLOR,
  ARTICLE_TITLE_COLOR
  // FEED_LINK_COLOR
} from "../../../layout/global/colors"

export const CodeWrapper = styled.div`
  margin-bottom: 1.8em;
`

const Link = styled.a`
  text-decoration: underline;
`
const ArticleWrapper = styled.div`
  font-size: 0.9em;
  color: ${ARTICLE_COLOR};
  line-height: 1.3em;
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 1em 0;
    color: ${ARTICLE_TITLE_COLOR};
  }
`

const CodeBlock = props => (
  <CodeWrapper>
    <PrismCodeBlock {...props} />
  </CodeWrapper>
)

const List = styled.ul`
  margin-bottom: 0.25em;
`
const ListItem = styled.li`
  margin-top: 0.25em;
`
const Paragraph = styled.p`
  margin-bottom: 0.8em;
`

const InlineCode = styled.span`
  border: 1px solid #ccc;
  background: #eee;
  border-radius: 4px;
  padding: 0.1em 0.4em;
  margin: 0.1em;
`

export const RawMarkdown: FC<{ markdown: string }> = ({ markdown }) => {
  return (
    <ArticleWrapper>
      <ReactMarkdown
        source={markdown}
        escapeHtml={false} // for web components
        renderers={{
          paragraph: Paragraph,
          code: CodeBlock,
          link: Link,
          list: List,
          listItem: ListItem,
          inlineCode: InlineCode
        }}
        plugins={[nl2brRemark]}
      />
    </ArticleWrapper>
  )
}
