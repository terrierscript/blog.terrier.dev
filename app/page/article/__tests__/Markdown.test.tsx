import React from "react"
import renderer from "react-test-renderer"
import { RawMarkdown, CodeBlock } from "../Markdown"

const removeIndent = str => str.replace(/\n\s+/gm, "\n")

test("Sample block", () => {
  const sampleMarkdown = removeIndent(`
    # Foo
    ## Baz
    ### bar
    foo baz bar
    - a
    - b
    - c
  `)
  const tree = renderer
    .create(<RawMarkdown markdown={sampleMarkdown} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

describe("Syntax Highlight", () => {
  const codeSample = `
    const foo = hoge
    const onClick = (e: Event) => {
      dispatch(e ** 2)
    }
    <SomeComponent>
      hello {foo}
    </SomeComponent>
  `
  const code = ["```tsx", codeSample, "â€œ```"].join("\n")
  const sampleMarkdown = removeIndent(`
    This is code
    ${code}
  `)
  it("as Markdown", () => {
    const tree = renderer
      .create(<RawMarkdown markdown={sampleMarkdown} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("as CodeBlock", () => {
    const tree = renderer
      .create(<CodeBlock value={codeSample} language={"tsx"} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
