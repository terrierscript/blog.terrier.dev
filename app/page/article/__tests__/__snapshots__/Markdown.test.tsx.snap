// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sample block 1`] = `
Array [
  <h1>
    Foo
  </h1>,
  <h2>
    Baz
  </h2>,
  <h3>
    bar
  </h3>,
  <p>
    foo baz bar
  </p>,
  <ul>
    <li>
      a
    </li>
    <li>
      b
    </li>
    <li>
      c
    </li>
  </ul>,
]
`;

exports[`Syntax Highlight as CodeBlock 1`] = `
<pre
  style={
    Object {
      "MozHyphens": "none",
      "MozTabSize": "4",
      "OTabSize": "4",
      "WebkitHyphens": "none",
      "background": "#2d2d2d",
      "color": "#ccc",
      "fontFamily": "Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace",
      "hyphens": "none",
      "lineHeight": "1.5",
      "margin": ".5em 0",
      "msHyphens": "none",
      "overflow": "auto",
      "padding": "1em",
      "tabSize": "4",
      "textAlign": "left",
      "whiteSpace": "pre",
      "wordBreak": "normal",
      "wordSpacing": "normal",
      "wordWrap": "normal",
    }
  }
>
  <code
    style={
      Object {
        "MozHyphens": "none",
        "MozTabSize": "4",
        "OTabSize": "4",
        "WebkitHyphens": "none",
        "background": "none",
        "color": "#ccc",
        "fontFamily": "Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace",
        "hyphens": "none",
        "lineHeight": "1.5",
        "msHyphens": "none",
        "tabSize": "4",
        "textAlign": "left",
        "whiteSpace": "pre",
        "wordBreak": "normal",
        "wordSpacing": "normal",
        "wordWrap": "normal",
      }
    }
  >
    
    
    <span
      className="token"
      style={
        Object {
          "color": "#cc99cd",
        }
      }
    >
      const
    </span>
     foo 
    <span
      className="token"
      style={
        Object {
          "color": "#67cdcc",
        }
      }
    >
      =
    </span>
     hoge
    
    <span
      className="token"
      style={
        Object {
          "color": "#cc99cd",
        }
      }
    >
      const
    </span>
     
    <span
      className="token function-variable"
      style={
        Object {
          "color": "#f08d49",
        }
      }
    >
      onClick
    </span>
     
    <span
      className="token"
      style={
        Object {
          "color": "#67cdcc",
        }
      }
    >
      =
    </span>
     
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      (
    </span>
    <span
      className="token parameter"
      style={Object {}}
    >
      e
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        :
      </span>
       
      <span
        className="token maybe-class-name"
        style={Object {}}
      >
        Event
      </span>
    </span>
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      )
    </span>
     
    <span
      className="token arrow"
      style={
        Object {
          "color": "#67cdcc",
        }
      }
    >
      =&gt;
    </span>
     
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      {
    </span>
    
      
    <span
      className="token"
      style={
        Object {
          "color": "#f08d49",
        }
      }
    >
      dispatch
    </span>
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      (
    </span>
    e 
    <span
      className="token"
      style={
        Object {
          "color": "#67cdcc",
        }
      }
    >
      **
    </span>
     
    <span
      className="token"
      style={
        Object {
          "color": "#f08d49",
        }
      }
    >
      2
    </span>
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      )
    </span>
    
    
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      }
    </span>
    
    
    <span
      className="token"
      style={
        Object {
          "color": "#e2777a",
        }
      }
    >
      <span
        className="token"
        style={
          Object {
            "color": "#e2777a",
          }
        }
      >
        <span
          className="token"
          style={
            Object {
              "color": "#ccc",
            }
          }
        >
          &lt;
        </span>
        <span
          className="token"
          style={
            Object {
              "color": "#f8c555",
            }
          }
        >
          SomeComponent
        </span>
      </span>
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        &gt;
      </span>
    </span>
    <span
      className="token plain-text"
      style={Object {}}
    >
      
      hello 
    </span>
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      {
    </span>
    foo
    <span
      className="token"
      style={
        Object {
          "color": "#ccc",
        }
      }
    >
      }
    </span>
    <span
      className="token plain-text"
      style={Object {}}
    >
      
    
    </span>
    <span
      className="token"
      style={
        Object {
          "color": "#e2777a",
        }
      }
    >
      <span
        className="token"
        style={
          Object {
            "color": "#e2777a",
          }
        }
      >
        <span
          className="token"
          style={
            Object {
              "color": "#ccc",
            }
          }
        >
          &lt;/
        </span>
        <span
          className="token"
          style={
            Object {
              "color": "#f8c555",
            }
          }
        >
          SomeComponent
        </span>
      </span>
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        &gt;
      </span>
    </span>
    
  
  </code>
</pre>
`;

exports[`Syntax Highlight as Markdown 1`] = `
Array [
  <p>
    This is code
  </p>,
  <pre
    style={
      Object {
        "MozHyphens": "none",
        "MozTabSize": "4",
        "OTabSize": "4",
        "WebkitHyphens": "none",
        "background": "#2d2d2d",
        "color": "#ccc",
        "fontFamily": "Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace",
        "hyphens": "none",
        "lineHeight": "1.5",
        "margin": ".5em 0",
        "msHyphens": "none",
        "overflow": "auto",
        "padding": "1em",
        "tabSize": "4",
        "textAlign": "left",
        "whiteSpace": "pre",
        "wordBreak": "normal",
        "wordSpacing": "normal",
        "wordWrap": "normal",
      }
    }
  >
    <code
      style={
        Object {
          "MozHyphens": "none",
          "MozTabSize": "4",
          "OTabSize": "4",
          "WebkitHyphens": "none",
          "background": "none",
          "color": "#ccc",
          "fontFamily": "Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace",
          "hyphens": "none",
          "lineHeight": "1.5",
          "msHyphens": "none",
          "tabSize": "4",
          "textAlign": "left",
          "whiteSpace": "pre",
          "wordBreak": "normal",
          "wordSpacing": "normal",
          "wordWrap": "normal",
        }
      }
    >
      <span
        className="token"
        style={
          Object {
            "color": "#cc99cd",
          }
        }
      >
        const
      </span>
       foo 
      <span
        className="token"
        style={
          Object {
            "color": "#67cdcc",
          }
        }
      >
        =
      </span>
       hoge

      <span
        className="token"
        style={
          Object {
            "color": "#cc99cd",
          }
        }
      >
        const
      </span>
       
      <span
        className="token function-variable"
        style={
          Object {
            "color": "#f08d49",
          }
        }
      >
        onClick
      </span>
       
      <span
        className="token"
        style={
          Object {
            "color": "#67cdcc",
          }
        }
      >
        =
      </span>
       
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        (
      </span>
      <span
        className="token parameter"
        style={Object {}}
      >
        e
        <span
          className="token"
          style={
            Object {
              "color": "#ccc",
            }
          }
        >
          :
        </span>
         
        <span
          className="token maybe-class-name"
          style={Object {}}
        >
          Event
        </span>
      </span>
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        )
      </span>
       
      <span
        className="token arrow"
        style={
          Object {
            "color": "#67cdcc",
          }
        }
      >
        =&gt;
      </span>
       
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        {
      </span>
      

      <span
        className="token"
        style={
          Object {
            "color": "#f08d49",
          }
        }
      >
        dispatch
      </span>
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        (
      </span>
      e 
      <span
        className="token"
        style={
          Object {
            "color": "#67cdcc",
          }
        }
      >
        **
      </span>
       
      <span
        className="token"
        style={
          Object {
            "color": "#f08d49",
          }
        }
      >
        2
      </span>
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        )
      </span>
      

      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        }
      </span>
      

      <span
        className="token"
        style={
          Object {
            "color": "#e2777a",
          }
        }
      >
        <span
          className="token"
          style={
            Object {
              "color": "#e2777a",
            }
          }
        >
          <span
            className="token"
            style={
              Object {
                "color": "#ccc",
              }
            }
          >
            &lt;
          </span>
          <span
            className="token"
            style={
              Object {
                "color": "#f8c555",
              }
            }
          >
            SomeComponent
          </span>
        </span>
        <span
          className="token"
          style={
            Object {
              "color": "#ccc",
            }
          }
        >
          &gt;
        </span>
      </span>
      <span
        className="token plain-text"
        style={Object {}}
      >
        
hello 
      </span>
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        {
      </span>
      foo
      <span
        className="token"
        style={
          Object {
            "color": "#ccc",
          }
        }
      >
        }
      </span>
      <span
        className="token plain-text"
        style={Object {}}
      >
        

      </span>
      <span
        className="token"
        style={
          Object {
            "color": "#e2777a",
          }
        }
      >
        <span
          className="token"
          style={
            Object {
              "color": "#e2777a",
            }
          }
        >
          <span
            className="token"
            style={
              Object {
                "color": "#ccc",
              }
            }
          >
            &lt;/
          </span>
          <span
            className="token"
            style={
              Object {
                "color": "#f8c555",
              }
            }
          >
            SomeComponent
          </span>
        </span>
        <span
          className="token"
          style={
            Object {
              "color": "#ccc",
            }
          }
        >
          &gt;
        </span>
      </span>
      
“
      <span
        className="token template-string"
        style={Object {}}
      >
        <span
          className="token template-punctuation"
          style={
            Object {
              "color": "#7ec699",
            }
          }
        >
          \`
        </span>
        <span
          className="token template-punctuation"
          style={
            Object {
              "color": "#7ec699",
            }
          }
        >
          \`
        </span>
      </span>
      \`
    </code>
  </pre>,
]
`;
